contracts/validator-set/ValidatorSet.sol

```diff
@@ -0,0 +1,294 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+/**
+ * @title ValidatorSet contract.
+ *
+ * @notice It contains methods to maintain validators for a metablock.
+ */
+contract ValidatorSet {
+
+    /* Usings */
+
+    using SafeMath for uint256;
+
+
+    /* Constants */
+
+    /** Maximum future end height, set for all active validators. */
+    uint256 public constant MAX_FUTURE_END_HEIGHT =
+        uint256(0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff);
+
+    /** Sentinel pointer for marking start and end of linked-list of validators. */
+    address public constant SENTINEL_VALIDATORS = address(0x1);
+
+
+    /* Storage */
+
+    /** Linked list of validators. */
+    mapping(address => address) public validators;
+
+    /**
+     * Validator begin height assigned to this set:
+     *   - zero: not registered to this set, or started at height 0
+     *           if endHeight > 0.
+     *   - bigger than zero: begin height for active validators.
+     */
+    mapping(address => uint256) public validatorBeginHeight;
+
+    /**
+     * Validator end height assigned to this set:
+     *   - zero: not registered to this set.
+     *   - MAX_FUTURE_END_HEIGHT: for active validators.
+     *   - less than MAX_FUTURE_END_HEIGHT: for logged out validators.
+     */
+    mapping(address => uint256) public validatorEndHeight;
+
+    /**
+     * Validator set count per height.
+     * Recursively the validator count, N, for height, h, can be written as:
+     *     N_h = N_(h-1) + J_h - L_(h-1)
+     * where J_h is the number of validators joining at height h, and
+     * L_(h-1) is the number of validators that logged out at h-1 (equals their end-height).
+     */
+    mapping(uint256 /* metablock height */ => uint256 /* validator count */) public validatorCount;
+
+    /**
+     * Forward validator set (FVS) count per height.
+     * The number of validators in the FVS at a given height, F_h, equals
+     * the number of validators at that height, N_h, minus the validators that have logged out at h:
+     *     F_h = N_h - L_h
+     */
+    mapping(uint256 /* metablock height */ => uint256 /* FVS count */) fvsCount;
+
+    /**
+     * Active height constrains the insertion and removal of validators
+     * to the active height. Any insertion or removal can increment
+     * the active height with one.
+     * Active height is private to ValidatorSet and intended to run ahead of
+     * the (open) metablock height from Core or Protocore with only a loose coupling.
+     * Active height is introduced to the calculation of the forward validator set count
+     * to a sliding-window calculation.
+     */
+    uint256 public activeHeight;
```

Comment:
As per the documentation of `activeHeight`, it must be :
`uint256 private activeHeight;`